Add NPR Rendering to GraphicsTools/StandardShader

HoloMoto opened the pull request:
## Overview

NPR functionality has been added to the light mode of GraphicsTools/Standard shaders.

Please see the following video for the actual operation.

https://youtu.be/rB6KaXwDAB4

The use of MeshOutline enables comic-like rendering.

　I would like to add new features to ToonShader in the future, but even at this point, we think it is possible to create some very interesting expressions.

　Personally, I think that the ability to use the existing StandardShader features in conjunction with the MeshOutline allows for a much greater variety of expressions!



## Changes
- Fixes: #108



## Verification

https://youtu.be/rB6KaXwDAB4

![image](https://user-images.githubusercontent.com/36409613/195098243-69716475-fa9f-4a2b-9a79-0816b9b36563.png)

## Additional
 @Cameron-Micka 
　Sorry I'm a little late as I was learning the new LightMode process!

　TokyoHoloLensMeetup will be held on 10/12 in Japan!
I will be presenting MRGT in a 5 minute slot. I'm off to talk about what MRGT has to offer, including this NPR feature!

Cameron-Micka commented on com.microsoft.mrtk.graphicstools.unity/Runtime/Shaders/GraphicsToolsLighting.hlsl:
```diff
@@ -1,6 +1,6 @@
 // Copyright � 2020 Unity Technologies ApS
 // Licensed under the Unity Companion License for Unity-dependent projects--see https://unity3d.com/legal/licenses/Unity_Companion_License
-// Unless expressly provided otherwise, the Software under this license is made available strictly on an �AS IS� BASIS WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED. Please review the license for details on these and other terms and conditions.
+// Unless expressly provided otherwise, the Software under this license is made available strictly on an �AS IS�EBASIS WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED. Please review the license for details on these and other terms and conditions.
```
Nitpick, looks like an 'E' was added.

```suggestion
// Unless expressly provided otherwise, the Software under this license is made available strictly on an AS IS BASIS WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED. Please review the license for details on these and other terms and conditions.

```

Cameron-Micka commented on com.microsoft.mrtk.graphicstools.unity/Editor/ShaderGUIs/StandardShaderGUI.cs:
```diff
@@ -110,9 +111,10 @@ protected static class Styles
             public static readonly GUIContent mipmapBias = new GUIContent("Mipmap Bias", "Degree to bias the mip map. A larger negative value reduces aliasing and improves clarity, but may decrease performance");
             public static readonly GUIContent lightMode = new GUIContent("Light Mode", "What Type of Direct Light Affects the Surface");
-            public static readonly string[] lightModeNames = new string[] { "Unlit", "Lit - Directional", "Lit - Distant" };
+            public static readonly string[] lightModeNames = new string[] { "Unlit", "Lit - Directional", "Lit - Distant" , "Non - Photorealistic" };
```
I love that you put this option in a drop down so that the UI doesn't become cluttered with too many checkboxes. But there might be scenarios where people want to use [DistantLights](https://github.com/microsoft/MixedReality-GraphicsTools-Unity/pull/37) with Non-Photorealistic rendering?

If that is the case maybe we move this property down to a separate check box that gets displayed when "Lit - Directional" or "Lit - Distant" is selected. Just like how specular highlights and spherical harmonics are toggled: 
![image](https://user-images.githubusercontent.com/13305729/195491970-5ac99ca9-ce3d-45a2-898d-3ec31969bc72.png)

Cameron-Micka commented on com.microsoft.mrtk.graphicstools.unity/Editor/ShaderGUIs/StandardShaderGUI.cs:
```diff
@@ -110,9 +111,10 @@ protected static class Styles
+            public static readonly string[] lightModeNames = new string[] { "Unlit", "Lit - Directional", "Lit - Distant" , "Non - Photorealistic" };
             public static readonly string lightModeLitDirectional = "_DIRECTIONAL_LIGHT";
             public static readonly string lightModeLitDistant = "_DISTANT_LIGHT";
+            public static readonly string lightModeNPR = "_NPR_Rendering";
```
nit: all caps to match the other keywords.

```suggestion
            public static readonly string lightModeNPR = "_NPR_RENDERING";

```

Cameron-Micka reviewed (commented):
This is amazing! What a wonderful idea and great execution. I added a few comments mostly about how to charge how the feature is enabled in the shader GUI. Once that is addressed this looks good to merge. 

Also had fun doing some testing:
![image](https://user-images.githubusercontent.com/13305729/195492636-2aa2d958-2a06-492d-8ad2-e6bfbdfa6e7f.png)

![image](https://user-images.githubusercontent.com/13305729/195492650-fd16bca4-e013-433d-94cd-a03f283eff85.png)

Also, if the Tokyo HoloLens Meetup was recored it would be fun to watch your segment.  😊

HoloMoto commented:
@Cameron-Micka 
I've incorporated the Corrections !
It certainly needed to be used in conjunction with DistanceLight!
Also fixed the overlap between NPRRendering and Rendering. (I need to get better acquainted with the English language...)
<img width="600" alt="20221016185943" src="https://user-images.githubusercontent.com/36409613/196029656-24f8e396-a705-4f7c-aec3-635ad9a71997.png">

Cameron-Micka commented on com.microsoft.mrtk.graphicstools.unity/Runtime/Shaders/GraphicsToolsStandard.shader:
```diff
@@ -34,6 +34,7 @@ Shader "Graphics Tools/Standard"
 
         // Rendering options.
         [Enum(Microsoft.MixedReality.GraphicsTools.Editor.LightMode)] _DirectionalLight("Light Mode", Float) = 1.0 // "LitDirectional"
+        [Toggle(_NON_PHOTOREALISTIC)] _NPR("Non-Photorealistic Rendering", Float) = 0.0
```
Could you also add this property to the `GraphicsToolsStandardCanvas` shader too? Without it you get this error in Unity:

> ArgumentException: Could not find MaterialProperty: '_NPR', Num properties: 117

Cameron-Micka commented on com.microsoft.mrtk.graphicstools.unity/Runtime/Shaders/GraphicsToolsLighting.hlsl:
```diff
@@ -18,7 +17,7 @@ struct GTMainLight
 {
     GTMainLight light;
-#if defined(_DIRECTIONAL_LIGHT) || defined(_DISTANT_LIGHT)
+#if defined(_DIRECTIONAL_LIGHT) || defined(_DISTANT_LIGHT)|| defined(_NON_PHOTOREALISTIC)
```
nitpick: You probably don't need this define?
```suggestion
#if defined(_DIRECTIONAL_LIGHT) || defined(_DISTANT_LIGHT)

```

Cameron-Micka reviewed (changes requested):
Tested the latest changes, thank you for responding to the feedback. It looks wonderful! This is good to merge after fixing the "Could not find MaterialProperty" exception and fixing the small merge conflicts. 😊

HoloMoto commented:
@Cameron-Micka 
Thank you!
I've corrected the two points you made!

Cameron-Micka reviewed (approved):
Looks great! This is ready to merge.
